File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function(head, left, right) {
    let prev = null;
    let curr = head;
    let cnt = 1;

    while (cnt < left) {
        prev = curr;
        curr = curr.next;
        cnt++;
    }

    let beforeReverse = prev;

    let reversedTail = curr;
    prev = null;
    while (cnt <= right) {
        let next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
        cnt++;
    }

    if (beforeReverse !== null) {
        beforeReverse.next = prev;
    } else {
        head = prev;
    }
    reversedTail.next = curr;

    return head;
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
     ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* prev = NULL;
        ListNode* curr = head;
        int cnt = 1;

        while (cnt < left) {
            prev = curr;
            curr = curr->next;
            cnt++;
        }

        ListNode* beforeReverse = prev;

        ListNode* reversedTail = curr;
        prev = NULL;
        while (cnt <= right) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
            cnt++;
        }

        if (beforeReverse != NULL) {
            beforeReverse->next = prev;
        } else {
            head = prev;
        }
        reversedTail->next = curr;

        return head;
    }
};